src/include/interval_set.h

```diff
@@ -56,20 +56,20 @@ class interval_set {
         }
 
         // Dereference this iterator to get a pair.
-        std::pair < T, T > &operator*() {
+        auto& operator*() const {
```

Comment:
i'd keep the type for better readability. the same applies to all the public interfaces.

so an alternative is 
```suggestion
        typename Map::reference operator*() const {
```
or
```suggestion
        std::pair<const T, T>& operator*() const {
```

Reply:
I'm have to disagree here. I don't think `Map::reference` is much easier to grasp than `auto`; you have to look up `std::map<T,T>::reference` and then determine it's really referring to `std::pair<const T, T>` which is, unfortunately, not something most developers are used to thinking about (i.e. the const-ness of the key).

Finally, these iterators are more an interface for the compiler and not really an interface for development. This is a heavily templated class so it's already difficult to understand; `auto` is supposed to help in these situations. An IDE helps with determining the type if necessary. (I don't personally use one.)

Reply:
> unfortunately, not something most developers are used to thinking about (i.e. the const-ness of the key).

that's exactly we need to document them in this way. no, i think they are interface for development.

Reply:
Alright. I've gone ahead and made your first suggested change. What do you think?